src/game/AI/ScriptDevAI/scripts/northrend/ulduar/ulduar/boss_xt_002.cpp

```diff
@@ -100,98 +102,96 @@ enum
+                SendAIEvent(AI_EVENT_CUSTOM_A, m_creature, heart);
 
-        m_uiLightBombTimer      = 10000;
-        m_uiGravityBombTimer    = 20000;
-        m_uiTanctrumTimer       = 35000;
+            DoBroadcastText(EMOTE_HEART, m_creature);
 
-        // reset flags and stand state
-        m_creature->SetStandState(UNIT_STAND_STATE_STAND);
-        m_creature->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_UNINTERACTIBLE);
+            m_creature->CastSpell(nullptr, SPELL_SUBMERGED, TRIGGERED_IGNORE_CURRENT_CASTED_SPELL | TRIGGERED_IGNORE_GCD | TRIGGERED_HIDE_CAST_IN_COMBAT_LOG | TRIGGERED_IGNORE_CASTER_AURA_STATE);
+        });
     }
 
-    void JustDied(Unit* /*pKiller*/) override
-    {
-        if (m_pInstance)
-            m_pInstance->SetData(TYPE_XT002, DONE);
+    instance_ulduar* m_instance;
+    bool m_isRegularMode;
 
-        DoScriptText(SAY_DEATH, m_creature);
-    }
+    uint8 m_heartStage;
 
-    void KilledUnit(Unit* pVictim) override
+    void Aggro(Unit* who) override
     {
-        if (pVictim->GetTypeId() != TYPEID_PLAYER)
-            return;
-
-        DoScriptText(urand(0, 1) ? SAY_SLAY_1 : SAY_SLAY_2, m_creature);
+        BossAI::Aggro(who);
+        if (m_instance)
+            m_instance->SetData(TYPE_XT002_HARD, NOT_STARTED);
     }
 
-    void Aggro(Unit* /*pWho*/) override
+    void EnterEvadeMode() override
     {
-        if (m_pInstance)
-        {
-            m_pInstance->SetData(TYPE_XT002, IN_PROGRESS);
-            m_pInstance->SetData(TYPE_XT002_HARD, NOT_STARTED);
-        }
-
-        DoScriptText(SAY_AGGRO, m_creature);
+        BossAI::EnterEvadeMode();
+        if (Creature* heart = m_instance->GetSingleCreatureFromStorage(NPC_HEART_DECONSTRUCTOR))
+            heart->ForcedDespawn();
     }
 
-    void JustReachedHome() override
+    void JustRespawned() override
     {
-        if (m_pInstance)
-        {
-            m_pInstance->SetData(TYPE_XT002, FAIL);
-
-            // mount the Heart back at the right seat after wipe or respawn (respawn handled in DB)
-            m_uiMountTimer = 1000;
-        }
+        StopTargeting(false);
+        SetCombatScriptStatus(false);
+        m_creature->SetStandState(UNIT_STAND_STATE_STAND);
+        m_heartStage = 1;
+        ResetTimer(XT002_MOUNT_HEART, 1s);
     }
 
-    void ReceiveAIEvent(AIEventType eventType, Unit* /*pSender*/, Unit* pInvoker, uint32 /*uiMiscValue*/) override
+    void ReceiveAIEvent(AIEventType eventType, Unit* /*sender*/, Unit* invoker, uint32 /*miscValue*/) override
     {
         // enable hard mode
-        if (eventType == AI_EVENT_CUSTOM_B && pInvoker->GetEntry() == NPC_HEART_DECONSTRUCTOR)
+        if (eventType == AI_EVENT_CUSTOM_B && invoker->GetEntry() == NPC_HEART_DECONSTRUCTOR)
         {
             // reset to normal phase and don't allow the boss to get back to heart phases
             DoResetToNormalPhase();
-            m_uiHeartStage = 4;
+            m_heartStage = 4;
```

Comment:
Might make sense to set that to 5 instead of 4, to 100% ensure there's no race condition when XT dies